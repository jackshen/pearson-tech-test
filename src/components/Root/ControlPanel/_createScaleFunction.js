export const EXPONENTIAL_DOMAIN = { max: 1, min: -1 };

// this function assumes a minimum of -1 and a maximum of 1
// offsetAmount represents how skewed the x intercept is from 0
// see the README for more information
export const createScaleFunction = (offsetAmount) => {
  // k and a are constants used for the curve calculation
  const k = Math.log(Math.sqrt((Math.exp(offsetAmount) + 1) / (Math.exp(offsetAmount) - 1)));
  const a = Math.sqrt((Math.exp(offsetAmount) + 1) * (Math.exp(offsetAmount) - 1));

  return {
    x: (x) => a * Math.exp(k * x) - Math.exp(offsetAmount),
    y: (y) => (1 / k) * Math.log((y + Math.exp(offsetAmount)) / a), // inverse of x
  };
};
