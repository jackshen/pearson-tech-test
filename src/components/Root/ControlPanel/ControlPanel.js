import React from "react";
import styled from "styled-components";

import { createScaleFunction, EXPONENTIAL_DOMAIN } from "./_createScaleFunction";

const OFFSET_AMOUNT = 0.5;
const SCALE_DEC_PLACES = 1;

const ControlPanelWrapper = styled.div`
  align-items: center;
  bottom: 3rem;
  display: flex;
  flex-direction: row;
  left: 50%;
  position: absolute;
  transform: translate(-50%, 0);
  width: max-content;
`;

const InputNumber = styled.input`
  border: 0.0625rem solid black;
  padding: 0.2rem 0.5rem;
  margin: 0 1rem;
`;

const thumbStyles = `
  appearance: none;
  background: rgb(217, 217, 217);
  border: 0.0625rem solid black;
  border-radius: 50%;
  cursor: pointer;
  height: 1rem;
  width: 1rem;
`;

const InputSlider = styled.input.attrs({ type: "range" })`
  appearance: none;
  height: 0.0625rem;
  background: black;
  width: 10rem;

  /* if we don't separate the two, it doesn't work */
  ::-webkit-slider-thumb {
    ${thumbStyles}
  }

  ::-moz-range-thumb {
    ${thumbStyles}
  }
`;

const scale = createScaleFunction(OFFSET_AMOUNT);

const ControlPanel = ({ onChange: pushChange, max, min, numSteps, value }) => {
  const halfInterval = (max - min) / 2;

  const handleChangeNumber = evt => {
    const xValue = Math.min(max, Math.max(min, Number(evt.target.value)));
    pushChange(xValue);
  };

  const handleChangeSlider = evt => {
    const yValue = evt.target.value;
    const xValue = scale.x(yValue) * halfInterval;
    pushChange(xValue);
  };

  const exponentialDomainRange = EXPONENTIAL_DOMAIN.max - EXPONENTIAL_DOMAIN.min;
  const valueRange = max - min;
  // value used in the inputSlider, a representation of value's relativity to min/max, but constrained between -1 and 1 (exponentialDomainRange)
  const inputSliderValue = EXPONENTIAL_DOMAIN.min + (exponentialDomainRange * (value - min)) / valueRange;

  return (
    <ControlPanelWrapper>
      {/* I purposely refrain from using htmlFor with an id, because the use of a static ID within a component makes multiple copies of the component semantically incorrect */}
      {/* I'm aware that there are ways around this issue; however, I have opted to simply nest the input within the label instead, maintaining both accessibility and semantics */}
      <label>
        Velocity (km/s)
        <InputNumber max={max} min={min} onChange={handleChangeNumber} type="number" value={value} />
      </label>
      <InputSlider
        data-testid="input-slider"
        onChange={handleChangeSlider}
        max={EXPONENTIAL_DOMAIN.max}
        min={EXPONENTIAL_DOMAIN.min}
        step={(EXPONENTIAL_DOMAIN.max - EXPONENTIAL_DOMAIN.min) / numSteps}
        value={scale.y(inputSliderValue)}
      />
    </ControlPanelWrapper>
  );
};

export default ControlPanel;
