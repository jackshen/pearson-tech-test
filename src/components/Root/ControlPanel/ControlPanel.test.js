import { fireEvent, render } from "@testing-library/react";
import React from "react";

import ControlPanel from "./ControlPanel";

describe("ControlPanel", () => {
  test("InputSlider works properly", async () => {
    const handleChangeMock = jest.fn();

    const { getByTestId } = render(
      <ControlPanel max={100} min={-100} numSteps={200} onChange={handleChangeMock} value={0} />
    );

    fireEvent.change(getByTestId("input-slider"), {
      target: { value: "0.8" }
    });

    expect(handleChangeMock.mock.calls[handleChangeMock.mock.calls.length - 1][0].toFixed(1)).toMatchSnapshot();

    fireEvent.change(getByTestId("input-slider"), {
      target: { value: "0.1" }
    });

    expect(handleChangeMock.mock.calls[handleChangeMock.mock.calls.length - 1][0].toFixed(1)).toMatchSnapshot();

    fireEvent.change(getByTestId("input-slider"), {
      target: { value: "-0.5" }
    });

    expect(handleChangeMock.mock.calls[handleChangeMock.mock.calls.length - 1][0].toFixed(1)).toMatchSnapshot();
  });
});
