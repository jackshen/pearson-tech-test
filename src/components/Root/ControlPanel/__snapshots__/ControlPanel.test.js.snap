// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanel InputSlider works properly 1`] = `"65.2"`;

exports[`ControlPanel InputSlider works properly 2`] = `"-24.2"`;

exports[`ControlPanel InputSlider works properly 3`] = `"-72.7"`;
